### ISSUE TITLE
[firebase_messaging]: Data only message not delivered to firebase onMessage listener in Foreground

### ISSUE THREAD

**Issue by thiagommohallem** [firebase/flutterfire#11688]
### Is there an existing issue for this?

- [X] I have searched the existing issues.

### Are you aware of the differences between iOS and Android background message handling?

- [X] I understand that iOS and Android background messages behave differently, and I've designed my application with that in mind.

### Do you have an active Apple Developer account?

- [X] I have an active Apple Developer account.

### Are you using a physical iOS device to test background messages?

- [X] I am using a physical iOS device to test background messages.

### Have you enabled "Remote Notifications" & "Background Mode" (Checking options for "Background Processing" & "Remote Notifications") in your app's Xcode project?

![Captura de Tela 2023-10-11 às 10 22 20](https://github.com/firebase/flutterfire/assets/36980264/baf388b7-02da-4bdf-a0e4-869439044662)

### Have you created an APNs key in your Apple Developer account & uploaded this APNs key to your Firebase console?

![Captura de Tela 2023-10-11 às 10 23 59](https://github.com/firebase/flutterfire/assets/36980264/7b85aa65-3945-4f58-b0f5-c1f93120a0c0)

### Have you disabled method swizzling for Firebase in your app?

![Captura de Tela 2023-10-11 às 10 25 05](https://github.com/firebase/flutterfire/assets/36980264/14cf08c7-b11d-4fa8-8098-51338d3e6253)

### Are you sending messages to your app from the Firebase Admin SDK?

```
admin
  .messaging()
  .send(
    
    {
      token: token,
      data: {
        type:'example',
      }
    },
  false,
  )
  .then((res) => {
    if (res.failureCount) {
      console.log('Failed', res.results[0].error);
    } else {
      console.log('Success');
    }
  })
  .catch((err) => {
    console.log('Error:', err);
  });
```

### Have you requested permission from the user to receive notifications?

- [X] I have the relevant permission to receive notifications.

### Have you used the 'Console' application on your macOS device to check if the iOS device's system is throttling your background messages?

```
padrão	10:30:55.722393-0300	SpringBoard	Received incoming message on topic br.com.pushapplication at priority 5
padrão	10:30:55.751090-0300	SpringBoard	[br.com.pushapplication] Received remote notification request B5D0-3548 [ waking: 0, hasAlertContent: 0, hasSound: 0 hasBadge: 0 hasContentAvailable: 1 hasMutableContent: 0 pushType: Background]
padrão	10:30:55.751188-0300	SpringBoard	[br.com.pushapplication] Process delivery of push notification B5D0-3548
padrão	10:30:55.751246-0300	SpringBoard	[br.com.pushapplication] Request DUET delivers content-available push notification to application
padrão	10:30:55.751513-0300	SpringBoard	SUBMITTING: com.apple.pushLaunch.br.com.pushapplication:F3EB4A
padrão	10:30:55.753089-0300	dasd	Submitted Activity: com.apple.pushLaunch.br.com.pushapplication:F3EB4A at priority 5 (Wed Oct 11 10:30:55 2023 - Thu Oct 12 10:30:55 2023)
padrão	10:30:55.761168-0300	dasd	com.apple.pushLaunch.br.com.pushapplication:F3EB4A:[
	{name: ApplicationPolicy, policyWeight: 50.000, response: {Decision: Absolutely Must Proceed, Score: 1.00, Rationale: [{[appIsForeground]: Required:1.00, Observed:1.00},]}}
 ], FinalDecision: Absolutely Must Proceed}

```

### Additional context and comments

I'm using Flutter version 3.10.4
Firebase Messaging version: 14.6.2

iOS device: iPhone 8
iOS version: 16.3.1

I'm sending a data only message to my Flutter app running in foreground using the admin script above. However, It's not being processed by onMessage listener, on android everything is fine. The only thing that appears on my log is this:

```
Warning: Application delegate received call to -application:didReceiveRemoteNotification:fetchCompletionHandler: but the completion handler was never called.
```

The message works well if the 'notification' part is present, but I need the push to be silent. Any help?

Here's my AppDelegate.swift:
![Captura de Tela 2023-10-11 às 10 38 47](https://github.com/firebase/flutterfire/assets/36980264/15fd9b17-c03d-4ca9-8f85-231feb3e9edf)

**Comment by darshankawar** [comment]
Thanks for the report @thiagommohallem 
Check [this](https://github.com/firebase/flutterfire/issues/11008) issue and see if it resembles your case and helps to resolve your case or not.

**Comment by thiagommohallem** [comment]
Thank you for the reply @darshankawar, but I'm afraid not. I only see the problem with data-only notifications exclusively on my iOS device. I'm able to send a 'normal' notification and trigger the onMessage method as shown below:
![Captura de Tela 2023-10-12 às 08 08 14](https://github.com/firebase/flutterfire/assets/36980264/769c25a2-b8d1-4ae6-a50a-694468177467)

Only the silent ones does not make through the listener, the only thing that prints out in the console is this:
![Captura de Tela 2023-10-12 às 08 08 56](https://github.com/firebase/flutterfire/assets/36980264/bd366a8a-533f-457e-87b0-d75c098ce7d8)

On Android everything works perfectly.

**Comment by darshankawar** [comment]
Thanks for the update @thiagommohallem 
You may check the [plugin example](https://github.com/firebase/flutterfire/tree/master/packages/firebase_messaging/firebase_messaging/example) and see its implementation and compare it with yours to check if any config is missing at your end or not.

**Comment by thiagommohallem** [comment]
@darshankawar I've replaced my app code with the example you sent me. Same results. Data-only messages are not captured by the onMessage listener with my app on the foreground on an iOS device. The only differences that I saw in native config files was that my AppDelegate is in Swift, the example is in Objective-c and that in the info.plist, processing is enabled for the UIBackrgoundModes, tried that as well.

Still got the same warning:
`Warning: Application delegate received call to -application:didReceiveRemoteNotification:fetchCompletionHandler: but the completion handler was never called.`

It seems to me that the silent-push method is not being handled natively for iOS

**Comment by thiagommohallem** [comment]
Updates: I was able to override the didReceiveRemoteNotification method in my AppDelegate.swift adding this code to invoke onMessage for data-only a.k.a silent push's
```
override func application(_ application: UIApplication, didReceiveRemoteNotification userInfo: [AnyHashable : Any], fetchCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Void) {
    
        let controller : FlutterViewController = window?.rootViewController as! FlutterViewController
        let channel = FlutterMethodChannel(name: "plugins.flutter.io/firebase_messaging", binaryMessenger: controller.binaryMessenger);
        
        let dictionary: NSDictionary = [
                    "messageId": "1697221363163183",
                    "data": [
                        "type":  "update-customer-wallet"
                    ],
                ]
        channel.invokeMethod("Messaging#onMessage", arguments: dictionary)
    }
```

Doing that, my push was received by the onMessage method as shown below:
![1](https://github.com/firebase/flutterfire/assets/36980264/c68d9444-e3dd-4224-9802-0142c871d169)

I think the firebase_messaging plugin is failing to call the channel method, or the didReceiveRemoteNotification implementation is not being called on the plugin side at all.

**Comment by darshankawar** [comment]
Thanks for the update @thiagommohallem 
Re: `didReceiveRemoteNotification`, check [this](https://github.com/firebase/flutterfire/issues/10082) issue with underlying comments for reference and see if they help in your case or not.

**Comment by thiagommohallem** [comment]
@darshankawar I was able to make my own implementation for the didReceiveRemoteNotification method to handle data-only messages in iOS. I've tried the example code and got the same errors results, I'd be great if you or anyone in the team could try it to see if this data-only foreground in iOS is a bug in my project/device or in the plugin.

I've put breakpoints everywhere during the 'investigation' part of this bug. The firebase messaging plugin here was not picking up anything from silent push's in foreground. But was working properly for notifications with alerts.

**Comment by yostrokon** [comment]
@thiagommohallem 
this works for ios data messaging (silent), the important thing is an option with content available

payload = {
      'data': {
          "click_action": "FLUTTER_NOTIFICATION_CLICK",
          "sound": "default",
          "priority": "high",
         "myData" : "myData"
      },
fcm.sendToDevice(tokens, payload, {'content_available': true})

**Comment by msarkrish** [comment]
> @thiagommohallem this works for ios data messaging (silent), the important thing is an option with content available
> 
> payload = { 'data': { "click_action": "FLUTTER_NOTIFICATION_CLICK", "sound": "default", "priority": "high", "myData" : "myData" }, fcm.sendToDevice(tokens, payload, {'content_available': true})

Hi @yostrokon 
Everything is ok, We are sending notification using api post method. I can add click_action, sound, priority in data map but where to add content_available as true? 

Is mydata key also needed?

**Comment by yostrokon** [comment]
@msarkrish 
i'm using firebase function for sending push notification, and there is a third parameter for options, content_available true will trigger onMessage listener on IOS when notification missing in payload,
mydata is an example key for your data to send with the data message,

**Comment by msarkrish** [comment]
Ok @yostrokon . But still, we are facing that warning. listener is not getting called.

**Comment by msarkrish** [comment]
> Updates: I was able to override the didReceiveRemoteNotification method in my AppDelegate.swift adding this code to invoke onMessage for data-only a.k.a silent push's
> 
> ```
> override func application(_ application: UIApplication, didReceiveRemoteNotification userInfo: [AnyHashable : Any], fetchCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Void) {
>     
>         let controller : FlutterViewController = window?.rootViewController as! FlutterViewController
>         let channel = FlutterMethodChannel(name: "plugins.flutter.io/firebase_messaging", binaryMessenger: controller.binaryMessenger);
>         
>         let dictionary: NSDictionary = [
>                     "messageId": "1697221363163183",
>                     "data": [
>                         "type":  "update-customer-wallet"
>                     ],
>                 ]
>         channel.invokeMethod("Messaging#onMessage", arguments: dictionary)
>     }
> ```
> 
> Doing that, my push was received by the onMessage method as shown below: ![1](https://user-images.githubusercontent.com/36980264/275009943-c68d9444-e3dd-4224-9802-0142c871d169.png)
> 
> I think the firebase_messaging plugin is failing to call the channel method, or the didReceiveRemoteNotification implementation is not being called on the plugin side at all.

Thank you @thiagommohallem . Now issue got resolved.

**Comment by thiagommohallem** [comment]
I was using the official FCM plugin from NodeJS to send my notifications, tried content available inside aps property and it didn't work. Ended up doing my own implementation overriding the didReceiveRemoteNotification in AppDelegate.swift based on FLTFirebaseMessagingPlugin.m.

It could be something specific to my project @darshankawar. But apparently, @msarkrish was facing the same issue. It is working for me this way, of course it is preferable if Firebase handled it, but due to business demand, I had to make it work first. Don't think I will keep doing tests because it is 'fixed' for me. Any updates on this i will post here. Thanks for the help
